File: app/src/main/java/pl/orbitemobile/wspolnoty/activities/home/HomePresenter.java
package pl.orbitemobile.wspolnoty.activities.home;

import pl.orbitemobile.wspolnoty.activities.article.ArticlePresenter;
import pl.orbitemobile.wspolnoty.activities.contact.ContactActivity;
import pl.orbitemobile.wspolnoty.activities.home.domain.HomeScreen;
import pl.orbitemobile.wspolnoty.activities.home.logic.TodaysMassCalculator;
import pl.orbitemobile.wspolnoty.activities.hours.HoursActivity;
import pl.orbitemobile.wspolnoty.R;
import pl.orbitemobile.wspolnoty.activities.article.ArticleActivity;
import pl.orbitemobile.wspolnoty.activities.home.logic.DialogBuilder;
import pl.orbitemobile.wspolnoty.activities.where.WhereActivity;
import pl.orbitemobile.wspolnoty.data.entities.Article;

import android.content.Context;
import android.content.Intent;
import android.view.MenuItem;

import io.reactivex.SingleObserver;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public class HomePresenter implements HomeContract.Presenter {
    
    private final HomeContract.View mView;
    
    private final SingleObserver<Article[]> mArticlesRemoteObserver = new ArticlesRemoteObserver();
    
    private CompositeDisposable mDisposable = new CompositeDisposable();
    
    private Context mContext;
    
    private HomeScreen mUseCase;
    
    HomePresenter(HomeContract.View view, Context context) {
        mView = view;
        mContext = context;
        mUseCase = new HomeScreen();
    }
    
    @Override
    public void onViewCreated() {
        setTodaysMassTimes();
    }
    
    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        int itemThatWasClickedId = item.getItemId();
        if (itemThatWasClickedId == R.id.about_app) {
            DialogBuilder builder = new DialogBuilder();
            builder.showAboutDialog(mContext);
            return true;
        }
        return false;
    }
    
    @Override
    public void onRetryClick() {
        mUseCase.getRemoteArticles().subscribe(mArticlesRemoteObserver);
        mView.showLoadingScreen();
    }
    
    @Override
    public void start() {
        mUseCase.getRemoteArticles().subscribe(mArticlesRemoteObserver);
    }
    
    @Override
    public void stop() {
        mDisposable.clear();
    }
    
    public void onHoursButtonClick() {
        startActivity(HoursActivity.class);
    }
    
    public void onWhereButtonClick() {
        startActivity(WhereActivity.class);
    }
    
    public void onContactButtonClick() {
        startActivity(ContactActivity.class);
    }
    
    public void onArticleClick(Article article) {
        Intent intent = new Intent(mContext, ArticleActivity.class);
        intent.putExtra(ArticlePresenter.ARTICLE_URL, article.getArticleUrl());
        intent.putExtra(ArticlePresenter.TITLE, article.getTitle());
        intent.putExtra(ArticlePresenter.IMG_URL, article.getImgUrl());
        
        mContext.startActivity(intent);
    }
    
    private void startActivity(Class nextActivityClass) {
        Intent intent = new Intent(mContext, nextActivityClass);
        mContext.startActivity(intent);
    }
    
    private void setTodaysMassTimes() {
        TodaysMassCalculator todaysMassCalculator = new TodaysMassCalculator();
        String todays_mass = todaysMassCalculator.calculateTodaysMass();
        mView.setTodaysMass(todays_mass);
    }
    
    private class ArticlesRemoteObserver implements SingleObserver<Article[]> {
        
        @Override
        public void onSubscribe(@NonNull final Disposable d) {
            mDisposable.add(d);
        }
        
        @Override
        public void onError(Throwable e) {
            mView.showErrorMessage();
            e.printStackTrace();
        }
        
        @Override
        public void onSuccess(@NonNull final Article[] articles) {
            mView.showArticles(articles);
        }
    }
}

File: app/src/main/java/pl/orbitemobile/wspolnoty/activities/home/HomePresenter.kt
/*
 * Copyright (c) 2017. All Rights Reserved. Michal Jankowski orbitemobile.pl
 */
package pl.orbitemobile.wspolnoty.activities.home

import android.os.Bundle
import android.os.PersistableBundle
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import pl.orbitemobile.wspolnoty.activities.article.ArticleActivity
import pl.orbitemobile.wspolnoty.activities.contact.ContactActivity
import pl.orbitemobile.wspolnoty.activities.home.domain.HomeUseCase
import pl.orbitemobile.wspolnoty.activities.hours.HoursActivity
import pl.orbitemobile.wspolnoty.activities.news.NewsActivity
import pl.orbitemobile.wspolnoty.activities.where.WhereActivity
import pl.orbitemobile.wspolnoty.activities.word.WordActivity
import pl.orbitemobile.wspolnoty.data.dto.ArticleDTO
import pl.orbitemobile.wspolnoty.utilities.ActivityLauncher
import pl.orbitemobile.wspolnoty.utilities.TodaysMassCalculator

class HomePresenter(override var disposable: CompositeDisposable? = CompositeDisposable(),
                    override var view: HomeContract.View? = null) : HomeContract.Presenter {

    private var useCase: HomeContract.UseCase = HomeUseCase()

    private var activityUtility = ActivityLauncher()

    override fun onViewAttached() {
        setTodaysMassTimes()
    }


    override fun onRetryClick() {
        if (isNetwork()) {
            useCase.remoteArticles.subscribe(this)
            view?.showLoadingScreen()
        } else {
            view?.showNetworkToast()
        }
    }

    override fun onStart(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
        useCase.remoteArticles.subscribe(this)
    }

    override fun onHoursButtonClick() =
            activityUtility.startActivity(view?.context, HoursActivity::class.java)


    override fun onWordButtonClick() = activityUtility.startActivity(view?.context, WordActivity::class.java)


    override fun onWhereButtonClick() = activityUtility.startActivity(view?.context, WhereActivity::class.java)


    override fun onContactButtonClick() = activityUtility.startActivity(view?.context, ContactActivity::class.java)


    override fun onNewsButtonClick() = activityUtility.startActivity(view?.context, NewsActivity::class.java)


    override fun onArticleClick(article: ArticleDTO) {
        val extra = hashMapOf(
                ArticleDTO.KEY.ARTICLE_URL to article.articleUrl,
                ArticleDTO.KEY.TITLE to article.title,
                ArticleDTO.KEY.IMG_URL to article.imgUrl)
        activityUtility.startActivity(view?.context, ArticleActivity::class.java, extra)
    }

    private fun setTodaysMassTimes() {
        val todaysMassCalculator = TodaysMassCalculator()
        val todays_mass = todaysMassCalculator.calculateTodaysMass()
        view?.setTodayMass(todays_mass)
    }

    override fun onError(e: Throwable) {
        view?.showErrorMessage()
        e.printStackTrace()
    }

    override fun onSuccess(articles: Array<ArticleDTO>) {
        view?.showArticles(articles)
    }

    override fun onSubscribe(d: Disposable) {
        disposable?.addAll(d)
    }
}
